src/Toml/Parser.hs

```diff
@@ -159,10 +161,19 @@ tableNameP = lexeme $ between (char '[') (char ']') keyP
 
 -- Values
 
+decimalP :: Parser Integer
+decimalP = mkInteger <$> decimalStringP
+  where
+    decimalStringP   = fst <$> match (some digitChar >> optional (try $ many _digitsP))
+    _digitsP         = char '_' >> some digitChar
```

Comment:
I'm not really sure that `try $ many smth` is an efficient parser. Looks like you can move `try` to `_digitsP`, like:

```haskell
try (char '_') *> some digitChar
```

Also, if you replace `some digitChar` with `many digitChar` you can remove `optional` from `decimalStringP`. I hope it works! And if it works, it will make the code cleaner.

Reply:
It needs to be this way to pass the test `1_2_3_ == 1`. If the `optional` is removed and the `try` is moved into `_digitsP` then you get `1_2_3_ == 123`. I don't really know what the correct behaviour should be here so happy to change it if the second one is valid?

Reply:
Oops, that was unfortunate. Yeah, `1_2_3_ == 1` test is invalid, it should be `123` so feel free to change this particular test 👌